# Assignment 3: Descriptive Statistics
## Measures of Central Tendency and Variability using Iris Dataset

**Name:** ___________ | **Class:** T.E. | **Roll No:** ___________

## Problem Statement
To compute summary statistics such as mean, median, minimum, maximum, and standard deviation for a dataset grouped by a categorical variable, and to display statistical details including percentiles for each category using Python.

## Theory
**Measures of Central Tendency:**
- Mean: Average value
- Median: Middle value (50th percentile)

**Measures of Variability:**
- Standard Deviation: Spread of data around mean
- Min/Max: Range of values

**Grouped Statistics:** Data grouped by categorical variables to analyze patterns within subgroups.

**Percentiles:** 25% (Q1), 50% (median), 75% (Q3) — help understand data distribution.

In [37]:
# Step 1: Import Required Libraries
import pandas as pd
import numpy as np
from sklearn.datasets import load_iris

print("Libraries imported successfully!")

Libraries imported successfully!


In [38]:
# Step 2: Load the Iris Dataset
iris = load_iris()
print(f"Feature names: {iris.feature_names}")
print(f"Target names: {iris.target_names}")
print(f"Dataset shape: {iris.data.shape}")

Feature names: ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
Target names: ['setosa' 'versicolor' 'virginica']
Dataset shape: (150, 4)


In [39]:
# Step 3: Convert Dataset into DataFrame
df = pd.DataFrame(iris.data, columns=iris.feature_names)
print("DataFrame created!")
df.head()

DataFrame created!


,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
0,5.1,3.5,1.4,0.2
1,4.9,3.0,1.4,0.2
2,4.7,3.2,1.3,0.2
3,4.6,3.1,1.5,0.2
4,5.0,3.6,1.4,0.2


In [40]:
# Step 4: Add Species as a Categorical Column
# Map numeric targets (0,1,2) to species names
species_map = {0: 'setosa', 1: 'versicolor', 2: 'virginica'}
df['species'] = pd.Categorical([species_map[t] for t in iris.target])

print("Species column added.")
print(f"Species distribution:{df['species'].value_counts()}")

Species column added.
Species distribution:species
setosa        50
versicolor    50
virginica     50
Name: count, dtype: int64


In [48]:
# Step 5: Rename Columns for Clarity
df.columns = ['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'species']
print("Columns renamed:")
print(df.columns.tolist())
df.head()

Columns renamed:
['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'species']


,sepal_length,sepal_width,petal_length,petal_width,species
0,5.1,3.5,1.4,0.2,setosa
1,4.9,3.0,1.4,0.2,setosa
2,4.7,3.2,1.3,0.2,setosa
3,4.6,3.1,1.5,0.2,setosa
4,5.0,3.6,1.4,0.2,setosa


In [42]:
# Step 6: Group Data by Species
grouped = df.groupby('species')
print(f"Groups: {list(grouped.groups.keys())}")
print(f"Samples per group:")
print(grouped.size())

Groups: ['setosa', 'versicolor', 'virginica']
Samples per group:
species
setosa        50
versicolor    50
virginica     50
dtype: int64


C:\Users\praka\AppData\Local\Temp\ipykernel_46160\979475216.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  grouped = df.groupby('species')


In [53]:
# Step 7: Compute Summary Statistics for Each Group
print("=" * 80)
for name, group in grouped:
    print(f"{'=' * 40}")
    print(f"  Species: {name.upper()}")
    print(f"{'=' * 40}")
    stats = group.loc[:,df.columns[:-1]].describe().loc[["50%","std","mean"]]
    # stats = group.iloc[:, :4].agg(['mean', 'median', 'min', 'max', 'std'])
    print(stats.round(3))
    print()

  Species: SETOSA
      sepal_length  sepal_width  petal_length  petal_width
50%          5.000        3.400         1.500        0.200
std          0.352        0.379         0.174        0.105
mean         5.006        3.428         1.462        0.246

  Species: VERSICOLOR
      sepal_length  sepal_width  petal_length  petal_width
50%          5.900        2.800          4.35        1.300
std          0.516        0.314          0.47        0.198
mean         5.936        2.770          4.26        1.326

  Species: VIRGINICA
      sepal_length  sepal_width  petal_length  petal_width
50%          6.500        3.000         5.550        2.000
std          0.636        0.322         0.552        0.275
mean         6.588        2.974         5.552        2.026



In [44]:
# Step 8: Get List of Unique Species
species_list = df['species'].unique().tolist()
print("Species list:", species_list)

Species list: ['setosa', 'versicolor', 'virginica']


In [45]:
# Step 9 & 10: Detailed Statistics Using describe() for Each Species
for species in species_list:
    subset = df[df['species'] == species].iloc[:, :4]
    print(f"{'=' * 60}")
    print(f"  DETAILED STATISTICS: {species.upper()}")
    print(f"{'=' * 60}")
    print(subset.describe(percentiles=[0.25, 0.5, 0.75]).round(3))

  DETAILED STATISTICS: SETOSA
       sepal_length  sepal_width  petal_length  petal_width
count        50.000       50.000        50.000       50.000
mean          5.006        3.428         1.462        0.246
std           0.352        0.379         0.174        0.105
min           4.300        2.300         1.000        0.100
25%           4.800        3.200         1.400        0.200
50%           5.000        3.400         1.500        0.200
75%           5.200        3.675         1.575        0.300
max           5.800        4.400         1.900        0.600
  DETAILED STATISTICS: VERSICOLOR
       sepal_length  sepal_width  petal_length  petal_width
count        50.000       50.000         50.00       50.000
mean          5.936        2.770          4.26        1.326
std           0.516        0.314          0.47        0.198
min           4.900        2.000          3.00        1.000
25%           5.600        2.525          4.00        1.200
50%           5.900        2.800    

In [46]:
# Step 11: Display Overall Summary
print("=" * 60)
print("  OVERALL COMPARISON OF SPECIES (MEAN VALUES)")
print("=" * 60)
overall = grouped.mean()
print(overall.round(3))

print(f"{'=' * 60}")
print("  OVERALL COMPARISON OF SPECIES (MEDIAN VALUES)")
print("=" * 60)
print(grouped.median().round(3))

print(f"{'=' * 60}")
print("  OVERALL COMPARISON OF SPECIES (STD DEVIATION)")
print("=" * 60)
print(grouped.std().round(3))

  OVERALL COMPARISON OF SPECIES (MEAN VALUES)
            sepal_length  sepal_width  petal_length  petal_width
species                                                         
setosa             5.006        3.428         1.462        0.246
versicolor         5.936        2.770         4.260        1.326
virginica          6.588        2.974         5.552        2.026
  OVERALL COMPARISON OF SPECIES (MEDIAN VALUES)
            sepal_length  sepal_width  petal_length  petal_width
species                                                         
setosa               5.0          3.4          1.50          0.2
versicolor           5.9          2.8          4.35          1.3
virginica            6.5          3.0          5.55          2.0
  OVERALL COMPARISON OF SPECIES (STD DEVIATION)
            sepal_length  sepal_width  petal_length  petal_width
species                                                         
setosa             0.352        0.379         0.174        0.105
versicolor   

## Dataset Description
- **Dataset:** Iris Dataset
- **Source:** sklearn.datasets (built-in)
- **Records:** 150 samples (50 per species)
- **Features:** Sepal Length, Sepal Width, Petal Length, Petal Width
- **Classes (Target):** Setosa, Versicolor, Virginica

## Conclusion
Descriptive statistics computed for the Iris dataset grouped by species. Mean, median, min, max, and standard deviation were calculated for each species group. Detailed percentiles (25%, 50%, 75%) were displayed using describe(). The statistics reveal clear differences across the three Iris species, particularly in petal dimensions.